src/js/actions/search/menucommands.js

```diff
@@ -0,0 +1,176 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+define(function (require, exports) {
+    "use strict";
+
+    var _ = require("lodash"),
+        Immutable = require("immutable"),
+        OS = require("adapter/os"),
+        keyUtil = require("js/util/key"),
+        strings = require("i18n!nls/strings");
+
+    var events = require("js/events");
+
+    var menuActions = require("js/actions/menu");
+
+    /**
+     * Get a localized label for the given menu entry ID
+     *
+     * @private
+     * @param {string} id
+     * @return {string|Object.<string, string>}
+     */
+    var _getLabelForEntry = function (id) {
+        var parts = id.split("."),
+            labels = strings.MENU,
+            path = "";
+
+        parts.forEach(function (part) {
+            if (labels[part] === undefined) {
+                path = null;
+                return false;
+            }
+
+            if (labels[part].$MENU) {
+                path += labels[part].$MENU + ">";
+                labels = labels[part];
+            } else {
+                path += labels[part];
+            }
+        });
+
+        return path;
+    };
+
+    var _getShortcut = function (fullShortcut) {
+        var modifierBits = fullShortcut.modifiers,
+            keyChar = fullShortcut.keyChar,
+            keyCode = fullShortcut.keyCode,
+            modifierStrings = strings.SEARCH.MODIFIERS,
+            shortcut = "";
+
+        if (modifierBits) {
+            var modifiers = keyUtil.bitsToModifiers(modifierBits);
+            
+            _.forEach(Object.keys(modifiers), function (key) {
+                if (modifiers[key]) {
+                    shortcut += modifierStrings[key.toUpperCase()] + "+";
+                }
+            });
+        }
+
+        if (keyChar) {
+            shortcut += keyChar.toString().toUpperCase();
+        }
+        if (keyCode) {
+            // Currently just have keys for F12 and escape
+            switch (keyCode) {
+            case OS.eventKeyCode.KEY_F12:
+                shortcut += "F12";
+                break;
+            case OS.eventKeyCode.ESCAPE:
+                shortcut += modifierStrings.ESCAPE;
+                break;
+            }
+        }
+
+        return " (" + shortcut + ")";
+    };
+    
+    /**
+     * Make list of recent documents info so search store can create search options
+     * 
+     * @private
+     * @return {Immutable.List.<object>}
+    */
+    var _menuCommandSearchOptions = function () {
+        var menuStore = this.flux.store("menu"),
+            menu = menuStore.getApplicationMenu(),
+            menuMap = menu.rootMap,
+            roots = menu.roots;
+
+        var menuCommands = Immutable.List();
```

Comment:
Start this with a regular JS array, push directly into the array, and then convert it to Immutable.List().

This code creates A LOT of garbage otherwise.